### PR TITLE
add e2e tts model: JETS

### PR DESCRIPTION
**Pull request by imdanboy**
Hi, thanks for your great works; espnet

I experimented the e2e tts model (JETS) recently following 
paper: https://arxiv.org/abs/2203.16852
code: https://github.com/imdanboy/jets
and get some reasonable results on ljspeech, kss dataset.

Most part of the code is based on fastspeech2 and hifi-gan of espnet. It would be honor if it can be implemented in espnet.
I think an alignment module, related to an text-speech alignment learning, would be helpful at least to simplify the training of non-autoregressive tts model such as fastspeech.

Thanks.

### CODE REVIEW

**Review by kan-bayashi — CHANGES_REQUESTED** [review]
Thank you for your contribution.
This is really cool :) 
I have some requests:
- Could you fix CI errors? https://github.com/espnet/espnet/runs/6424148362?check_suite_focus=true 
- Could you add unit test for JETS modules 
  E.g., https://github.com/espnet/espnet/blob/afa8f8ec5b8ec77deb1a3c1531915ebbee7b80e6/test/espnet2/gan_tts/vits/test_vits.py#L280-L343

https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#41-python
This document will help you.

File: espnet2/gan_tts/jets/generator.py [hunk 1]
```diff
@@ -0,0 +1,795 @@
+# Copyright 2021 Tomoki Hayashi
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+"""Generator module in JETS.
+
+"""
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/generator.py:6]:
```suggestion
"""Generator module in JETS."""
```

File: espnet2/gan_tts/jets/loss.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        blank_prob: float = np.e**-1,
+    ) -> torch.Tensor:
+        """Calculate forward propagation.
+
+        Args:
+            log_p_attn (Tensor): Batch of log probability of attention matrix
+                (B, T_feats, T_text).
+            ilens (Tensor): Batch of the lengths of each input (B,).
+            olens (Tensor): Batch of the lengths of each target (B,).
+            blank_prob (float): Blank symbol probability
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/loss.py:136]:
```suggestion
            blank_prob (float): Blank symbol probability.
```

File: espnet2/gan_tts/jets/length_regulator.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+        """Upsample hidden states according to durations.
+
+        Args:
+            hs (Tensor): Batched hidden state to be expanded (B, T_text, adim)
+            ds (Tensor): Batched token duration (B, T_text)
+            h_masks (Tensor): Mask tensor (B,T_feats)
+            d_masks (Tensor): Mask tensor (B,T_text)
+
+        Returns:
+            Tensor: Expanded hidden state (B, T_feat, adim)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/length_regulator.py:27]:
```suggestion
            hs (Tensor): Batched hidden state to be expanded (B, T_text, adim).
            ds (Tensor): Batched token duration (B, T_text).
            h_masks (Tensor): Mask tensor (B, T_feats).
            d_masks (Tensor): Mask tensor (B, T_text).

        Returns:
            Tensor: Expanded hidden state (B, T_feat, adim).
```

File: espnet2/gan_tts/jets/alignments.py [hunk 4]
```diff
@@ -0,0 +1,163 @@
+import numpy as np
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/alignments.py:1]: Add your credit?

File: espnet2/gan_tts/jets/generator.py [hunk 5]
```diff
@@ -0,0 +1,795 @@
+# Copyright 2021 Tomoki Hayashi
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/generator.py:1]: Change to your credit?

File: espnet2/gan_tts/jets/jets.py [hunk 6]
```diff
@@ -0,0 +1,652 @@
+# Copyright 2021 Tomoki Hayashi
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/jets.py:1]: Change to your credit?

File: espnet2/gan_tts/jets/length_regulator.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import logging
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/length_regulator.py:1]: Add your credit?

**Review by kan-bayashi — COMMENTED** [review]
Thank you so much your kind modification.
Almost good for me but could you check my minor comments?

**Comment by imdanboy** [comment]
> Thank you so much your kind modification. Almost good for me but could you check my minor comments?

Thank you for your kind reply!
I've reflected your comments :)

File: test/espnet2/gan_tts/jets/test_jets.py [hunk 8]
```diff
@@ -0,0 +1,1024 @@
+# Copyright 2021 Tomoki Hayashi
```
Comment (kan-bayashi) [inline test/espnet2/gan_tts/jets/test_jets.py:1]:
```suggestion
# Copyright 2022 Dan Lim
```

File: espnet2/gan_tts/jets/loss.py [hunk 9]
```diff
@@ -0,0 +1,212 @@
+# Copyright 2020 Nagoya University (Tomoki Hayashi)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/loss.py:1]:
```suggestion
# Copyright 2022 Dan Lim
```

File: espnet2/gan_tts/jets/alignments.py [hunk 10]
```diff
@@ -0,0 +1,166 @@
...
+    def forward(self, text, feats, x_masks=None):
+        """Calculate alignment loss.
+
+        Args:
+            text (Tensor): Batched text embedding (B, T_text, adim)
+            feats (Tensor): Batched acoustic feature (B, T_feats, odim)
+            x_masks (Tensor): Mask tensor (B, T_text)
+
+        Returns:
+            Tensor: log probability of attention matrix (B, T_feats, T_text)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/alignments.py:37]:
```suggestion
            text (Tensor): Batched text embedding (B, T_text, adim).
            feats (Tensor): Batched acoustic feature (B, T_feats, odim).
            x_masks (Tensor): Mask tensor (B, T_text).

        Returns:
            Tensor: Log probability of attention matrix (B, T_feats, T_text).
```

File: espnet2/gan_tts/jets/alignments.py [hunk 11]
```diff
@@ -0,0 +1,166 @@
...
+
+    Args:
+        log_p_attn (Tensor): Batched log probability of attention
+            matrix (B, T_feats, T_text)
+        text_lengths (Tensor): Text length tensor (B,)
+        feats_legnths (Tensor): Feature length tensor (B,)
+
+    Returns:
+        Tensor: Batched token duration extracted from `log_p_attn` (B,T_text)
+        Tensor: binarization loss tensor ()
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/alignments.py:107]:
```suggestion
        log_p_attn (Tensor): Batched log probability of attention
            matrix (B, T_feats, T_text).
        text_lengths (Tensor): Text length tensor (B,).
        feats_legnths (Tensor): Feature length tensor (B,).

    Returns:
        Tensor: Batched token duration extracted from `log_p_attn` (B, T_text).
        Tensor: Binarization loss tensor ().
```

File: espnet2/gan_tts/jets/alignments.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+    """Average frame-level features into token-level according to durations
+
+    Args:
+        ds (Tensor): Batched token duration (B,T_text)
+        xs (Tensor): Batched feature sequences to be averaged (B,T_feats)
+        text_lengths (Tensor): Text length tensor (B,)
+        feats_lengths (Tensor): Feature length tensor (B,)
+
+    Returns:
+        Tensor: Batched feature averaged according to the token duration (B, T_text)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/alignments.py:158]:
```suggestion
        ds (Tensor): Batched token duration (B, T_text).
        xs (Tensor): Batched feature sequences to be averaged (B, T_feats).
        text_lengths (Tensor): Text length tensor (B,).
        feats_lengths (Tensor): Feature length tensor (B,).

    Returns:
        Tensor: Batched feature averaged according to the token duration (B, T_text).
```

File: espnet2/gan_tts/jets/generator.py [hunk 13]
```diff
@@ -0,0 +1,793 @@
...
+            Tensor: Waveform tensor (B, 1, segment_size * upsample_factor).
+            Tensor: binarization loss ()
+            Tensor: log probability attention matrix (B,T_feats,T_text)
+            Tensor: Segments start index tensor (B,).
+            Tensor: predicted duration (B,T_text)
+            Tensor: ground-truth duration obtained from an alignment module (B,T_text)
+            Tensor: predicted pitch (B,T_text,1)
+            Tensor: ground-truth averaged pitch (B,T_text,1)
+            Tensor: predicted energy (B,T_text,1)
+            Tensor: ground-truth averaged energy (B,T_text,1)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/generator.py:554]:
```suggestion
            Tensor: Waveform tensor (B, 1, segment_size * upsample_factor).
            Tensor: Binarization loss ().
            Tensor: Log probability attention matrix (B, T_feats, T_text).
            Tensor: Segments start index tensor (B,).
            Tensor: predicted duration (B, T_text).
            Tensor: ground-truth duration obtained from an alignment module (B, T_text).
            Tensor: predicted pitch (B, T_text,1).
            Tensor: ground-truth averaged pitch (B, T_text, 1).
            Tensor: predicted energy (B, T_text, 1).
            Tensor: ground-truth averaged energy (B, T_text, 1).
```

File: espnet2/gan_tts/jets/loss.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+    def _generate_prior(self, text_lengths, feats_lengths, w=1) -> torch.Tensor:
+        """Generate alignment prior formulated as beta-binomial distribution
+
+        Args:
+            text_lengths (Tensor): Batch of the lengths of each input (B,).
+            feats_lengths (Tensor): Batch of the lengths of each target (B,).
+            w (float): Scaling factor; lower -> wider the width
+
+        Returns:
+            Tensor: Batched 2d static prior matrix (B, T_feats, T_text)
```
Comment (kan-bayashi) [inline espnet2/gan_tts/jets/loss.py:183]:
```suggestion
            w (float): Scaling factor; lower -> wider the width.

        Returns:
            Tensor: Batched 2d static prior matrix (B, T_feats, T_text).
```

**Review by kan-bayashi — APPROVED** [review]
Thank you for your modification!
I sightly fixed minor issues.
If the CI is passed, I will merge this great PR :)

**Comment by imdanboy** [comment]
@sw005320, Sure! I will do your suggestion soon :)

**Comment by kan-bayashi** [comment]
@imdanboy I updated README #4369.
If you can the pretrained models, that is great!

**Pull request merged**